File: src/Heq/Storage/BandDto.cs
using System;
using Heq.Model;

namespace Heq.Storage
{
    public class BandDto
    {
        public FilterKind Type { get; set; }
        public double Freq { get; set; }
        public double Gain { get; set; }
        public double Q { get; set; }
        public int Slope { get; set; } = 12;
        public bool Enabled { get; set; } = true;
        public ChannelTarget Channel { get; set; }

        public static BandDto From(EqBand b) => new BandDto
        {
            Type = b.Kind,
            Freq = b.Freq,
            Gain = b.GainDb,
            Q = b.Q,
            Slope = b.SlopeDbPerOct,
            Enabled = b.Enabled,
            Channel = b.Channel,
        };

        public EqBand ToBand() => new EqBand
        {
            Kind = Type,
            Freq = Freq,
            GainDb = Gain,
            Q = Q,
            SlopeDbPerOct = Slope <= 0 ? 12 : Slope,
            Enabled = Enabled,
            Channel = Channel,
        };

        public bool Matches(EqBand b)
            => b.Kind == Type
            && b.Enabled == Enabled
            && b.Channel == Channel
            && b.SlopeDbPerOct == (Slope <= 0 ? 12 : Slope)
            && Same(b.Freq, Freq)
            && Same(b.GainDb, Gain)
            && Same(b.Q, Q);

        private static bool Same(double a, double b) => Math.Abs(a - b) <= 1e-9;
    }
}

File: rust/heq-core/src/storage/band_dto.rs
use serde::{Deserialize, Serialize};

use crate::model::{BandId, ChannelTarget, EqBand, FilterKind};

// the one place the stored band shape is defined; field names are the C# app's
#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
#[serde(rename_all = "PascalCase")]
pub struct BandDto {
    #[serde(rename = "Type")]
    pub kind: FilterKind,
    pub freq: f64,
    pub gain: f64,
    pub q: f64,
    #[serde(default = "default_slope")]
    pub slope: i32,
    #[serde(default = "default_enabled")]
    pub enabled: bool,
    #[serde(default)]
    pub channel: ChannelTarget,
}

fn default_slope() -> i32 {
    12
}

fn default_enabled() -> bool {
    true
}

impl BandDto {
    pub fn from_band(b: &EqBand) -> Self {
        BandDto {
            kind: b.kind,
            freq: b.freq,
            gain: b.gain_db,
            q: b.q,
            slope: b.slope_db_per_oct,
            enabled: b.enabled,
            channel: b.channel,
        }
    }

    pub fn to_band(self) -> EqBand {
        let mut b = EqBand::new(BandId(0));
        b.kind = self.kind;
        b.freq = self.freq;
        b.gain_db = self.gain;
        b.q = self.q;
        b.slope_db_per_oct = if self.slope <= 0 { 12 } else { self.slope };
        b.enabled = self.enabled;
        b.channel = self.channel;
        b.clamp();
        b
    }

    pub fn matches(&self, b: &EqBand) -> bool {
        b.kind == self.kind
            && b.enabled == self.enabled
            && b.channel == self.channel
            && b.slope_db_per_oct == if self.slope <= 0 { 12 } else { self.slope }
            && same(b.freq, self.freq)
            && same(b.gain_db, self.gain)
            && same(b.q, self.q)
    }
}

fn same(a: f64, b: f64) -> bool {
    (a - b).abs() <= 1e-9
}
